docs/reactive-patterns.qmd

```diff
@@ -0,0 +1,232 @@
+
+@reactive.calc
+def df():
+    # req() stops execution until input.file() is truthy
+    f = req(input.file())
+    return pd.read_csv(f[0]['datapath'])
+
+@render.data_frame
+def table():
+    # Output won't render until input.file() is truthy
+    return render.DataGrid(df())
+```
+:::
+
+![A gif of uploading a csv file and seeing the text and table outputs update](assets/file-upload.mp4){class="img-fluid shadow rounded" style="margin-top: -1.5rem;"}
+
+
+## Scheduled invalidation {#invalidate-later}
+
+Sometimes its useful invalidate a reactive function on a schedule.
+For example, to implement a clock that updates every second, you can use `reactive.invalidate_later(1)`:
+
+```{shinylive-python}
+#| standalone: true
+#| components: [editor, viewer]
+#| layout: vertical
+#| viewerHeight: 50
+from datetime import datetime
+from shiny import reactive
+from shiny.express import render
+
+# Get the current time every second
+@reactive.calc
+def cur_time():
+    reactive.invalidate_later(1)
+    return datetime.now().strftime('%H:%M:%S')
+
+@render.ui
+def clock():
+    return f"Current time: {cur_time()}"
+```
+
+
+## Reactive file reading {#file}
+
+Sometimes its useful invalidate a reactive function when a file changes.
+For example, lets extend the example from above to write the current time to a file every second, and then read and display the contents of that file:
+
+```{shinylive-python}
+#| standalone: true
+#| components: [editor, viewer]
+#| layout: vertical
+#| viewerHeight: 50
+from datetime import datetime
+from shiny import reactive
+from shiny.express import render
+
+# Get the current time every second
+@reactive.calc
+def cur_time():
+    reactive.invalidate_later(1)
+    return datetime.now().strftime('%H:%M:%S')
+
+# Write cur_time() to a file (every second)
+@reactive.effect
+def _():
+    with open("time.txt", "w") as f:
+        f.write(cur_time())
+
+f = open("time.txt", "w")  # Create the file if it doesn't exist
+
+# Read and display whenever the file changes
+@render.ui
+@reactive.file_reader("time.txt")
+def time():
+    with open("time.txt") as f:
+        return f"Current time {f.read()}"
+```
+
+## Reactive polling {#poll}
+
+Sometimes it's useful to invalidate a reactive function on a schedule, but only _if a certain condition is met_.
+For example, suppose we want to check if a (potentially large) file (or database) has changed/updated every so often, and if it has, re-read it.
+The `@reactive.poll()` decorator is designed for this purpose.
+When applying the decorator, make sure to provide a function that is relatively cheap to execute, since it will be executed repeatedly on an interval.
+And, in the event that that function's value changes, the reactive function will be invalidated and re-executed.
+
+For example, lets extend the example from above to write the current time to a file every 0.5 seconds, but only read and display the contents every 2 seconds:
```

Comment:
As above I think it's better to have a database query which generates a table and a plot, even if we have to tell instead of show what it does.